internal/querynode/task_search.go

```diff
@@ -0,0 +1,325 @@
+	OrigTopKs             []int64
+	Ret                   *internalpb.SearchResults
+	originTasks           []*searchTask
+}
+
+func (s *searchTask) PreExecute(ctx context.Context) error {
+	topK := s.TopK
+	if topK <= 0 || topK >= 16385 {
+		return fmt.Errorf("limit should be in range [1, 16385], but got %d", topK)
+	}
+	s.combinePlaceHolderGroups()
+	return nil
+}
+
+func (s *searchTask) searchOnStreaming() error {
+	// deserialize query plan
+	// check if collection has been released
+	collection, err := s.QS.streaming.replica.getCollectionByID(s.CollectionID)
+	if err != nil {
+		return err
+	}
+
+	searchReq, err2 := newSearchRequest(collection, s.req, s.PlaceholderGroup)
+	if err2 != nil {
+		return err2
+	}
+	defer searchReq.delete()
+
+	// TODO add context
+	partResults, _, _, sErr := s.QS.streaming.search(searchReq, s.CollectionID, s.iReq.GetPartitionIDs(), s.req.GetDmlChannel())
+
+	if sErr != nil {
+		log.Warn("failed to search streaming data", zap.Int64("collectionID", s.CollectionID), zap.Error(sErr))
+		return sErr
+	}
+	defer deleteSearchResults(partResults)
+	return s.reduceResults(searchReq, partResults)
+}
+
+func (s *searchTask) searchOnHistorical() error {
+	// check ctx timeout
+	if !funcutil.CheckCtxValid(s.Ctx()) {
+		return errors.New("search context timeout")
+	}
+	segmentIDs := s.req.GetSegmentIDs()
+
+	// lock historic meta-replica
+	s.QS.historical.replica.queryRLock()
+	defer s.QS.historical.replica.queryRUnlock()
+
+	// check if collection has been released
+	collection, err := s.QS.historical.replica.getCollectionByID(s.CollectionID)
+	if err != nil {
+		return err
+	}
+	if s.GuaranteeTimestamp >= collection.getReleaseTime() {
+		log.Warn("collection release before search", zap.Int64("collectionID", s.CollectionID))
+		return fmt.Errorf("retrieve failed, collection has been released, collectionID = %d", s.CollectionID)
+	}
+
+	searchReq, err2 := newSearchRequest(collection, s.req, s.PlaceholderGroup)
+	if err2 != nil {
+		return err2
+	}
+	defer searchReq.delete()
+	err = s.QS.historical.validateSegmentIDs(segmentIDs, s.CollectionID, s.iReq.GetPartitionIDs())
+	if err != nil {
+		log.Warn("segmentIDs in search request fails validation", zap.Int64s("segmentIDs", segmentIDs))
+		return err
+	}
+
+	partResults, _, err := s.QS.historical.searchSegments(searchReq, segmentIDs)
+	if err != nil {
+		return err
+	}
+	defer deleteSearchResults(partResults)
+	return s.reduceResults(searchReq, partResults)
+}
+
+func (s *searchTask) Execute(ctx context.Context) error {
+	if s.DataScope == querypb.DataScope_Streaming {
+		log.Info("searchTask Execute DataScope Streaming")
+		return s.searchOnStreaming()
+	} else if s.DataScope == querypb.DataScope_Historical {
+		log.Info("searchTask Execute DataScope Historical")
+		return s.searchOnHistorical()
+	}
+	panic("searchTask not implement search on all data scope")
+}
+
+func (s *searchTask) Notify(err error) {
+	s.done <- err
+	for i := 1; i < len(s.originTasks); i++ {
```

Comment:
Notify start from index 1 but not 0?

Reply:
originTasks[0] is the searchTask iteself.
In order not to cause infinite recursion, number 0 is skipped here.
s.done <- err in line 146 will noitfy  originTasks[0].